Optionally restrict the range of ephemeral ports

sirf opened the pull request:
Optionally restrict the range of ephemeral ports that the Connection class can bind to.
This is a proposed fix for https://github.com/aiortc/aioice/issues/47

sirf commented:
`asyncio.create_datagram_endpoint` is used in ice.py, mdns.py, and turn.py, but ice.py is the only one that binds to an ephemeral port. mdns.py binds to the mDNS port 5353, and turn.py has a non-optional argument to set the port, so I believe ice.py is the only place where a change is needed unless I'm missing something.
 
 
I'll try to put some tests together.

jlaine commented:
>  mdns.py binds to the mDNS port 5353,

You're right, we can forget about mDNS.

>  turn.py has a non-optional argument to set the port

I'm not sure what you mean here, we are not setting the local port:

https://github.com/aiortc/aioice/blob/2aed57f1fa0ffeedc3a5f257a8fbbace07ca2c1f/src/aioice/turn.py#L416-L425

Concerning the actual API, instead of `port_lower` and `port_upper` maybe have you considered `ephemeral_ports: Optional[Iterable[int]] = None` ? The docstring could also be more explicit about the fact these are *local* ports.

sirf commented:
> > turn.py has a non-optional argument to set the port
> 
> I'm not sure what you mean here, we are not setting the local port:
> 
> https://github.com/aiortc/aioice/blob/2aed57f1fa0ffeedc3a5f257a8fbbace07ca2c1f/src/aioice/turn.py#L416-L425

turn.py creates a socket that is associated with a known remote address/port, and for this reason I'm not sure if it's important to be able to restrict which local port to use here, but I'll admit that I'm not 100% sure. I moved the critical code to utils.py however, so if it is indeed desirable to be able to restrict the local port for STUN as well it should take me about 5 minutes to do that.

Do you have any input on this, [gregvish](https://github.com/gregvish) ?

> Concerning the actual API, instead of `port_lower` and `port_upper` maybe have you considered `ephemeral_ports: Optional[Iterable[int]] = None` ? The docstring could also be more explicit about the fact these are _local_ ports.

Good idea, that would allow a set of ports that is not restricted to a continuous sequence.
I've updated the code, and also added unit tests.

marshallmassengill commented:
I have a use for this with broadcasting data over a restricted port range.  Any chance this could get merged?

dawgster commented:
Merge would be very appreciated
